railties/lib/rails/application/finisher.rb

```diff
@@ -175,6 +175,7 @@ def self.complete(_state)
 
           routes_reloader.run_after_load_paths = -> do
             app.routes.append do
+              get "/favicon.ico" => "rails/welcome#favicon", internal: true
```

Comment:
We could also let the route handle the 204 and leave out the changes to the controller.
```suggestion
              get "/favicon.ico" => lambda { |_| [204, {}, []] }, internal: true
```
Or even:
```suggestion
              get "/favicon.ico" => lambda { |_| [200, {}, File.new('path/to/favicon')] } internal: true
```

Reply:
If would be nice if we could just do:
```
get "/favicon.ico" => :head, internal: true`
```

Reply:
Looking at the guides, they use the following syntax:
```
get "/favicon.ico", to: ->(env) { [204, {}, ['']] }, internal: true
```

Reply:
I considered using `get "/favicon.ico", to: ->(env) { [204, {}, ['']] }, internal: true`.

I ended up going with defining it on the controller, similar to the welcome controller because a new user is one day going to start digging into how the welcome controller is displayed and they will see it's just a basic route 😀

Consistency here is nice, but I'm more than happy to change it.

Reply:
The inline env has my preference as it is about as long as the "controller action" implementation, without an extra step of indirection.
But I don't have a strong opinion about it 😄